### DISCUSSION TITLE
Integrate swell-js with next-auth

### DISCUSSION THREAD

**Discussion by dudusotero** [swellstores/community#240] [Help]
Hello everyone!

Someone knows how we can integrate correctly swell-js with next-auth?
I'm creating a `CredentialsProvider` to handle it, but I can't intercept the `X-session` header used at `swell.account.login()` and append it to my jwt callback on next-auth.

Someone already was able to fully integrate it? With authentication, locale and currency values?

Here's my current implementation, can someone point me what am I missing?

```ts
export const authOptions: AuthOptions = {
  providers: [
    CredentialsProvider({
      name: 'Credentials',
      credentials: {
        email: { label: 'Email address', type: 'email' },
        password: { label: 'Password', type: 'password' },
      },
      async authorize(credentials): Promise<any> {
        const data = await swell.account.login(
          credentials?.email!,
          credentials?.password!
        )

        if (data) {
          return data
        } else {
          return null
        }
      },
    }),
  ],
  callbacks: {
    async jwt({ token }): Promise<any> {
      if (token) {
        const response = await swellFetch(`/accounts/${token.email}`, {
          method: 'PUT',
          body: JSON.stringify({ password_token: null }),
        })
        const data = await response.json()
        return data
      }

      return token
    },
    async session({ session }) {
      return {
        ...session,
        user: {
          name: session.user?.name,
          email: session.user?.email,
          image: `https://gravatar.com/avatar/${crypto
            .createHash('md5')
            .update(session.user?.email || '')
            .digest('hex')}`,
        },
      }
    },
  },
  session: {
    strategy: 'jwt',
  },
  pages: {
    signIn: '/signin',
  },
}
```

PS: the main idea is to make my user be able to sign in and see protected endpoints like `orders`, `profile` and even redirect to `checkout` already signed in.

Reference: https://developers.swell.is/storefronts/token-login

**Comment by rodrigoddalmeida** [comment]
Hi, @dudusotero - I see you have a `swellFetch` helper function that probably fetches directly from our REST Frontend API. You could use that for `/login`, so you have better control over the response headers and can extract/store `X-Session` for future requests. 

An alternative is to read the `swell-session` cookies, as they are updated accordingly when using Swell.js functions.

One thing you should note is that our session is stateful, meaning it can change after specific requests (like changing locale, currency for a user, changing carts, etc...), so you should keep it in sync with the latest value returned by the `X-Session` header.

**Reply by dudusotero** [reply]
Hi @rodrigoddalmeida, just to give you a little bit of context:

```ts
export function swellFetch(
  input: URL | RequestInfo,
  init?: RequestInit | undefined
): Promise<Response> {
  return fetch(`${SWELL_API_ENDPOINT}/${input}`, {
    ...init,
    headers: {
      ...init?.headers,
      Authorization: `Basic ${Buffer.from(
        `${NEXT_PUBLIC_SWELL_STORE_ID}:${SWELL_SECRET_KEY}`,
        'binary'
      ).toString('base64')}`,
    },
  })
}
```

I will try later today your suggestion, I will keep this discussion updated!
Thank you!

**Reply by dudusotero** [reply]
@rodrigoddalmeida about getting the `swell-session` cookie value, I don't know if will be possible, since all my requests are made into a SSR approach, and the cookie will not be injected on the client.

**Reply by rodrigoddalmeida** [reply]
@dudusotero Here's a reference to swell-js code handling the X-Session header: https://github.com/swellstores/swell-js/blob/25f787cc72dcfe9a9088266d6d0614266d17de2e/src/api.js#L172

I believe you can implement something similar, that you could then use to store a Swell session in a server session within your app.

**Reply by dudusotero** [reply]
@rodrigoddalmeida can you send me an example how the endpoint should be triggered? or an API doc for this particular `/login` endpoint.
What I'm currently have:
```ts
const data = await swellFetch('/login', {
  method: 'POST',
  body: JSON.stringify({
    email: credentials?.email,
    password: credentials?.password,
  }),
})
```

**Reply by rodrigoddalmeida** [reply]
Here's a reference from Swell.js: https://github.com/swellstores/swell-js/blob/25f787cc72dcfe9a9088266d6d0614266d17de2e/src/account.js#L25-L36

**Comment by dudusotero** [comment]
Hello everyone!

I decided to give up on trying to integrate it with next-auth, and implemented a custom solution with Next `middleware` + `useSWR` package.

Here's the ending result:

middleware.ts:
--
```ts
import type { NextRequest } from 'next/server'
import { NextResponse } from 'next/server'

const Authorization = `Basic ${Buffer.from(
  process.env.NEXT_PUBLIC_SWELL_PUBLIC_KEY!
).toString('base64')}`

export async function middleware(request: NextRequest) {
  const reqHeaders: HeadersInit = {
    Accept: 'application/json',
    'Content-Type': 'application/json',
    Authorization,
  }

  const session = request.cookies.get('swell-session')
  const locale = request.cookies.get('swell-locale')
  const currency = request.cookies.get('swell-currency')

  if (session) {
    reqHeaders['X-Session'] = session.value
  }

  if (locale) {
    reqHeaders['X-Locale'] = locale.value
  }

  if (currency) {
    reqHeaders['X-Currency'] = currency.value
  }

  const redirectUrl = new URL('/signin', request.url)
  redirectUrl.searchParams.set('redirectTo', request.nextUrl.href)
  const response = NextResponse.redirect(redirectUrl)

  try {
    const swellResponse = await fetch(
      `${process.env.SWELL_API_ENDPOINT}/account`,
      {
        method: 'get',
        headers: reqHeaders,
        credentials: 'include',
        mode: 'cors',
      }
    )

    const account = await swellResponse.json()

    if (!account) {
      return response
    }

    return NextResponse.next()
  } catch (error) {
    console.error(error)
    response.cookies.delete('swell-session')
  }

  return response
}

export const config = {
  // change it based on your scenario
  matcher: ['/profile/:path*'],
}
```
useUser.ts
--
```ts
import { getCurrentUser } from '@/lib/swell/account'
import useSWR from 'swr'

export default function useUser() {
  const { data, error, isLoading, isValidating, mutate } =
    useSWR<swell.Account | null>('/api/me', getCurrentUser)

  return {
    user: data,
    error,
    isLoading,
    isValidating,
    mutate,
  }
}
```

getCurrentUser()
--
```ts
export const getCurrentUser = () => swell.account.get()
```

In a near future I will come back to my old problem and will dive deep into how to integrate it with `next-auth`.
Thanks @rodrigoddalmeida for all the assistance.
